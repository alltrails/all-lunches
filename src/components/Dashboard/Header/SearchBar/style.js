import styled from 'styled-components';
import { colors } from 'styles';
import TransparentButton from 'styles/components/TransparentButton';

export const Form = styled.form`
  border: 1px solid ${colors.borderGray};
  box-shadow: 0px 1px 2px rgba(0, 0, 0, 0.1);
  border-radius: 6px;
  display: flex;
  flex-direction: row;
  align-items: center;
  overflow: hidden;
  width: 302px;
`;

export const TextInput = styled.input`
  outline: 0;
  border: none;
  width: 100%;
  height: 100%;
  font-size: 1.1rem;
  padding-left: 8px;

  &::placeholder {
    color: ${colors.grayDark};
    font-weight: bold;
  }
`;

export const Button = styled(TransparentButton)`
  padding: 12px;
  display: flex;
  align-items: center;
  justify-content: center;
  border: 1px solid transparent;
  position: relative;

  ${(p) =>
    !p.isLoading &&
    `
    &.focus-visible,
    &:hover {
      border: 1px solid ${colors.green};
      border-radius: 6px;
      padding: 4px 12px;
    }
    `}
`;

export const MagnifyingGlassImg = styled.img`
  height: 20px;
`;

export const LoadingWrapper = styled.div`
  position: absolute;
  left: 0;
  right: 0;
  top: 50%;
  transform: translateY(-50%);
`;
