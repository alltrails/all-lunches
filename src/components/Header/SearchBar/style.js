import styled from 'styled-components';
import { colors } from 'styles';
import TransparentButton from 'styles/components/TransparentButton';

export const Form = styled.form`
  border: 1px solid ${colors.borderGray};
  box-shadow: 0px 1px 2px rgba(0, 0, 0, 0.1);
  border-radius: 6px;
  display: flex;
  flex-direction: row;
  align-items: center;
  overflow: hidden;
  width: 302px;
`;

export const TextInput = styled.input`
  outline: 0;
  border: none;
  width: 100%;
  height: 100%;
  font-size: 1.1rem;
  padding-left: 8px;

  &::placeholder {
    color: ${colors.grayDark};
    font-weight: bold;
  }
`;

export const Button = styled(TransparentButton)`
  padding: 12px;
  height: 100%;
  display: flex;
  align-items: center;
  justify-content: center;
`;

export const MagnifyingGlassImg = styled.img`
  height: 20px;
`;
