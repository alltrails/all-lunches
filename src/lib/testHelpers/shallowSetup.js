import React from 'react';
import { shallow } from 'enzyme';

export default (Component, defaultProps) =>
  (otherProps = {}) => {
    const props = { ...defaultProps, ...otherProps };
    const wrapper = shallow(<Component {...props} />);

    return { props, wrapper };
  };
