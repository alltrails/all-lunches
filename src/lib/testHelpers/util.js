/* eslint-env jest */

export const createChangeEvent = (name, value) => ({
  target: { name, value },
});
