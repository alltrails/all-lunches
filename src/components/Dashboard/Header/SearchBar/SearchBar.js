import React from 'react';
import PropTypes from 'prop-types';

import magnifyingGlass from 'assets/magnifying-glass.svg';
import LoadingSpinner from 'components/shared/LoadingSpinner';

import { LoadingWrapper, MagnifyingGlassImg, Button, Form, TextInput } from './style';

const SearchBar = ({ isLoading, onSubmit, searchText, onChange }) => (
  <Form onSubmit={onSubmit}>
    <TextInput
      type="text"
      onChange={onChange}
      placeholder="Search for a restaurant"
      value={searchText}
    />
    <Button isLoading={isLoading} disabled={isLoading} onClick={onSubmit}>
      {isLoading ? (
        <LoadingWrapper>
          <LoadingSpinner />
        </LoadingWrapper>
      ) : (
        <MagnifyingGlassImg src={magnifyingGlass} alt="Search here" />
      )}
    </Button>
  </Form>
);

SearchBar.propTypes = {
  onSubmit: PropTypes.func.isRequired,
  onChange: PropTypes.func.isRequired,
  searchText: PropTypes.string.isRequired,
  isLoading: PropTypes.bool.isRequired,
};

export default SearchBar;
