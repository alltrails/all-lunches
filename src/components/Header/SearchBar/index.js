import React from 'react';

import magnifyingGlass from 'assets/magnifying-glass.svg';
import { MagnifyingGlassImg, Button, Form, TextInput } from './style';

const SearchBar = () => {
  console.log('test');

  return (
    <Form onSubmit={() => {}}>
      <TextInput type="text" onChange={() => {}} placeholder="Search for a restaurant" value="" />
      <Button>
        <MagnifyingGlassImg src={magnifyingGlass} alt="Search here" />
      </Button>
    </Form>
  );
};

export default SearchBar;
