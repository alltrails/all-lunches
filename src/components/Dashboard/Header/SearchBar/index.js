import React, { useState } from 'react';
import PropTypes from 'prop-types';
import { connect } from 'react-redux';

import { isLoadingSelector } from 'store/restaurants/selectors';
import * as restaurantsActions from 'store/restaurants/actions';

import SearchBar from './SearchBar';

const mapStateToProps = (state) => ({
  isLoading: isLoadingSelector(state),
});

const mapDispatchToProps = {
  queryArea: restaurantsActions.queryArea,
};

export const SearchBarContainer = ({ isLoading, queryArea }) => {
  const [searchText, setSearchText] = useState('');

  const handleChange = (event) => {
    const { value } = event.target;
    setSearchText(value);
  };

  const handleSubmit = (event) => {
    event.preventDefault();
    const searchTextTrimmed = searchText.trim();

    if (searchTextTrimmed) queryArea(searchTextTrimmed);
  };

  return (
    <SearchBar
      isLoading={isLoading}
      onSubmit={handleSubmit}
      onChange={handleChange}
      searchText={searchText}
    />
  );
};

SearchBarContainer.propTypes = {
  isLoading: PropTypes.bool.isRequired,
  queryArea: PropTypes.func.isRequired,
};

export default connect(mapStateToProps, mapDispatchToProps)(SearchBarContainer);
