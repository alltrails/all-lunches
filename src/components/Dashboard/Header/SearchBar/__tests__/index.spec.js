/* eslint-env jest */

import shallowSetup from 'lib/testHelpers/shallowSetup';
import { createChangeEvent } from 'lib/testHelpers/util';

import SearchBar from '../SearchBar';
import { SearchBarContainer } from '../index';

describe('SearchBarContainer', () => {
  const defaultProps = {
    isLoading: false,
    queryArea: () => {},
  };
  const setup = shallowSetup(SearchBarContainer, defaultProps);
  const searchInputText = 'coffee';

  it('renders the expected view', () => {
    const { wrapper } = setup();
    expect(wrapper).toMatchSnapshot();
  });

  it('renders the expected view when loading', () => {
    const { wrapper } = setup({ isLoading: true });
    expect(wrapper).toMatchSnapshot();
  });

  describe('Handling input change', () => {
    it('updates the searchText prop of the child component', () => {
      const { wrapper } = setup();
      let searchBarComponent = wrapper.find(SearchBar);

      expect(searchBarComponent.prop('searchText')).toEqual('');

      searchBarComponent.props().onChange(createChangeEvent('searchText', searchInputText));
      searchBarComponent = wrapper.find(SearchBar);

      expect(searchBarComponent.prop('searchText')).toEqual(searchInputText);
    });
  });

  describe('Handling submit', () => {
    it("prevents the submit event's default behavior", () => {
      const { wrapper } = setup();
      const submitEvent = { preventDefault: jest.fn() };
      const searchBarComponent = wrapper.find(SearchBar);

      searchBarComponent.props().onSubmit(submitEvent);

      expect(submitEvent.preventDefault).toBeCalled();
    });

    it('calls the queryArea callback prop with the searchText value', () => {
      const { props, wrapper } = setup({
        queryArea: jest.fn(),
      });
      const submitEvent = { preventDefault: () => {} };
      let searchBarComponent = wrapper.find(SearchBar);
      searchBarComponent.props().onChange(createChangeEvent('searchText', searchInputText));

      searchBarComponent = wrapper.find(SearchBar);
      searchBarComponent.props().onSubmit(submitEvent);

      expect(props.queryArea).toBeCalledWith(searchInputText);
    });

    it('does not call the queryArea callback prop with an input value of empty spaces', () => {
      const { props, wrapper } = setup({
        queryArea: jest.fn(),
      });
      const submitEvent = { preventDefault: () => {} };
      let searchBarComponent = wrapper.find(SearchBar);
      searchBarComponent.props().onChange(createChangeEvent('searchText', '    '));

      searchBarComponent = wrapper.find(SearchBar);
      searchBarComponent.props().onSubmit(submitEvent);

      expect(props.queryArea).not.toBeCalled();
    });
  });
});
