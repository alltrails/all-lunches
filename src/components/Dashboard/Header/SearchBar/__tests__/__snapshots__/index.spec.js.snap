// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SearchBarContainer renders the expected view 1`] = `
<SearchBar
  isLoading={false}
  onChange={[Function]}
  onSubmit={[Function]}
  searchText=""
/>
`;

exports[`SearchBarContainer renders the expected view when loading 1`] = `
<SearchBar
  isLoading={true}
  onChange={[Function]}
  onSubmit={[Function]}
  searchText=""
/>
`;
